wfc_gh/WaveFunctionCollapse.cs

```diff
@@ -394,141 +297,190 @@ protected override void SolveInstance(IGH_DataAccess DA)
                             return;
                     }
                 }
+            }
+
+            unsafe {
+                // Clear all modules to false first
+                for (ushort z = 0; z < worldZ; ++z) {
+                    for (ushort y = 0; y < worldY; ++y) {
+                        for (ushort x = 0; x < worldX; ++x) {
+                            for (ushort m = 0; m < nameToModule.Count; ++m) {
+                                var result = Native.wfc_world_state_slot_module_set(wfcWorldStateHandle, x, y, z, m, 0);
+                                switch (result) {
+                                    case WfcWorldStateSlotModuleSetResult.Ok:
+                                        break;
+                                    case WfcWorldStateSlotModuleSetResult.ErrSlotOutOfBounds:
+                                    case WfcWorldStateSlotModuleSetResult.ErrModuleOutOfBounds:
+                                    default:
+                                        Debug.Assert(false);
+                                        return;
+                                }
+                            }
+                        }
+                    }
+                }
+
+                // Set modules provided by user
+                for (var i = 0; i < worldSlotMinCount; ++i) {
+                    var position = worldSlotPositions[i];
+                    var moduleStr = worldSlotModules[i];
+
+                    int xInt = (int)Math.Round(position.X);
+                    int yInt = (int)Math.Round(position.Y);
+                    int zInt = (int)Math.Round(position.Z);
+
+                    if (xInt < 0 || yInt < 0 || zInt < 0) {
+                        AddRuntimeMessage(GH_RuntimeMessageLevel.Error,
+                                          "Slot positions must be nonnegative integers");
+
+                        Native.wfc_world_state_free(wfcWorldStateHandle);
+                        return;
+                    }
+
+                    ushort x = (ushort)xInt;
+                    ushort y = (ushort)yInt;
+                    ushort z = (ushort)zInt;
+
+                    if (nameToModule.TryGetValue(moduleStr.ToString(), out ushort m)) {
+                        var result = Native.wfc_world_state_slot_module_set(wfcWorldStateHandle, x, y, z, m, 1);
+                        switch (result) {
+                            case WfcWorldStateSlotModuleSetResult.Ok:
+                                break;
+                            case WfcWorldStateSlotModuleSetResult.ErrSlotOutOfBounds:
+                                AddRuntimeMessage(GH_RuntimeMessageLevel.Error,
+                                                  "Provided slots must be within bounds of the world");
+
+                                Native.wfc_world_state_free(wfcWorldStateHandle);
+                                return;
+                            case WfcWorldStateSlotModuleSetResult.ErrModuleOutOfBounds:
```

Comment:
does `ErrModuleOutOfBounds` say that the `m` value is larger than the module count? the error name is confusing

Reply:
Yes, that's exactly what it says. Suggestions?

Reply:
Yeah `out of bounds` is the usual wording. I just didn't realize it is about the module **index** and fort I was thinking of 3D coordinates.